File: gek-core/src/main/java/xyz/fslabo/common/proxy/JdkProxyProvider.java
package xyz.fslabo.common.proxy;

import xyz.fslabo.annotations.Nullable;
import xyz.fslabo.common.base.Jie;
import xyz.fslabo.common.coll.JieColl;
import xyz.fslabo.common.invoke.Invoker;
import xyz.fslabo.common.invoke.InvokingException;

import java.lang.reflect.*;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * JDK implementation for {@link ProxyProvider}, based on {@link Proxy}, only supports interface proxy.
 *
 * @author fredsuvn
 */
public class JdkProxyProvider implements ProxyProvider {

    private static final Object[] EMPTY_ARGS = {};

    @Override
    public <T> T newProxyInstance(@Nullable ClassLoader loader, Iterable<Class<?>> proxied, MethodProxyHandler handler) {
        ClassLoader actualLoader = loader == null ? getClass().getClassLoader() : loader;
        Object proxy = Proxy.newProxyInstance(actualLoader, JieColl.toArray(proxied, Class.class), new InvocationHandlerImpl(proxied, handler));
        return Jie.as(proxy);
    }

    private static final class InvocationHandlerImpl implements InvocationHandler {

        private final MethodProxyHandler handler;
        private final Set<Method> proxiedSet = new HashSet<>();
        private final Map<Method, Invoker> invokerMap = new HashMap<>();

        private InvocationHandlerImpl(Iterable<Class<?>> uppers, MethodProxyHandler handler) {
            this.handler = handler;
            for (Class<?> upper : uppers) {
                Method[] methods = upper.getMethods();
                for (Method method : methods) {
                    if (Modifier.isFinal(method.getModifiers()) || Modifier.isStatic(method.getModifiers())) {
                        continue;
                    }
                    invokerMap.put(method, Invoker.handle(method));
                    if (handler.proxy(method)) {
                        proxiedSet.add(method);
                    }
                }
            }
        }

        @Override
        public Object invoke(Object proxy, Method method, @Nullable Object[] args) throws Throwable {
            Invoker invoker = invokerMap.get(method);
            if (!proxiedSet.contains(method)) {
                return invoke0(invoker, proxy, args);
            }
            return handler.invoke(proxy, method, args, new ProxyInvoker() {
                @Override
                public @Nullable Object invoke(Object inst, Object[] args) throws Throwable {
                    return invoke0(invoker, inst, args);
                }

                @Override
                public @Nullable Object invokeSuper(Object[] args) throws Throwable {
                    throw new AbstractMethodError(method.toString());
                }
            });
        }

        private Object invoke0(Invoker invoker, Object inst, @Nullable Object[] args) throws Throwable {
            try {
                Object[] actualArgs = args == null ? EMPTY_ARGS : args;
                return invoker.invoke(inst, actualArgs);
            } catch (InvokingException e) {
                Throwable cause = e.getCause();
                if (cause instanceof InvocationTargetException) {
                    throw cause.getCause();
                }
                throw cause;
            }
        }
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/proxy/JdkProxyProvider.kt
package xyz.srclab.common.proxy

import xyz.srclab.common.base.asType
import xyz.srclab.common.invoke.StaticInvoke
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy

/**
 * JDK implementation for [ClassProxyProvider].
 */
object JdkProxyProvider : ClassProxyProvider {

    override fun <T : Any> getProxy(
        sourceClass: Class<T>,
        proxyInvoker: ProxyInvoker,
        classLoader: ClassLoader,
    ): ClassProxy<T> {
        return ProxyClassImpl(sourceClass, proxyInvoker, classLoader)
    }

    private class ProxyClassImpl<T : Any>(
        private val sourceClass: Class<T>,
        private val proxyInvoker: ProxyInvoker,
        private val classLoader: ClassLoader,
    ) : ClassProxy<T> {

        override fun newInst(): T {
            return Proxy.newProxyInstance(
                classLoader,
                arrayOf(sourceClass),
                ProxyMethodInterceptor(proxyInvoker)
            ).asType()
        }

        override fun newInst(parameterTypes: Array<Class<*>>, args: Array<Any?>): T {
            throw IllegalArgumentException("JDK proxy only supports interface.")
        }
    }

    private class ProxyMethodInterceptor(
        private val proxyInvoker: ProxyInvoker,
    ) : InvocationHandler {

        override fun invoke(proxy: Any, method: Method, args: Array<out Any>?): Any? {
            if (proxyInvoker.isTarget(method)) {
                return proxyInvoker.invokeProxy(proxy, method, UnsupportedSourceInvoke, args)
            }
            throw IllegalStateException("Proxy method not found: $method")
        }
    }

    private object UnsupportedSourceInvoke : StaticInvoke {
        override fun invoke(vararg args: Any?): Any? {
            throw IllegalStateException("Interface's method is not implemented.")
        }
    }
}